MudFileUpload: Improve validation in MudForm

igotinfected opened the pull request:
Builds upon #7578 to flesh out form validation with `MudFileUpload`, allowing users to clear a `MudFileUpload` instance by setting `Files` to `null`. 

## Description

We have a form where users can upload images for a specific model. When **editing** an existing item, if all the user intends to do is clear the existing image on the form page, the `MudForm` instance will not be marked as `Touched`.

Visually we have a `Clear` button that programmatically sets `Files` to `null`, very similar to the drag and drop example in the [docs](https://mudblazor.com/components/fileupload#drag-and-drop-example).

[Here's](https://try.mudblazor.com/snippet/cOGdlkxlVEqbmqBE) a snippet -- pressing the clear button will remove the file but not trigger `IsTouched = true`.

Unsure if the proposed solution makes sense, though I somewhat based it on the `MudDatePicker` implementation.

## How Has This Been Tested?
bUnit tested

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

igotinfected committed:
MudFileUpload: Allow clearing FileUpload by setting `Files` to null

Builds upon #7578 to flesh out form validation on `MudFileUpload` by
allowing users to clear a `MudFileUpload` instance by setting `Files` to
`null`. This would previously not trigger form validation or it would
not mark the `MudFileUpload` instance as `Touched`.

ScarletKuro commented:
Adding @Mr-Technician, but I'm not really liking that we are adding an async call here in components parameter.
I know we have such pattern in MudBlazor's codebase, but we are trying to get rid of it and I would like to avoid it in the new code(this can lead to infinity render issue https://github.com/dotnet/aspnetcore/issues/28518 especially that your property set is calling an FilesChanged depending on the codepath).
But I will let @Mr-Technician to decide it.

Mr-Technician commented:
@ScarletKuro I was looking at this last night and had the same thought. I don't want to add an async call in the setter.

@igotinfected How about implementing a Clear button? This is how other Form components do it.

igotinfected commented:
@Mr-Technician @ScarletKuro Fully agree with both of you guys, this was the quick and dirty version :)

I think for a clean `Clear` button approach we would have to make a breaking change on `ButtonTemplate` to provide a `Clear` callback via the context, similar to how Blazorise does it:

```csharp
<ButtonsTemplate>
    <Progress Value="@filePickerCustom.GetProgressPercentage()" />
    <Buttons>
        <Button Clicked="@context.Clear" Color="Color.Warning"><Icon Name="IconName.Clear" /></Button>
        <Button Clicked="@context.Upload" Color="Color.Primary"><Icon Name="IconName.FileUpload" /></Button>
    </Buttons>
</ButtonsTemplate>
```

We would then have something akin to:

```csharp
public RenderFragment<FileUploadButtonTemplateContext> ButtonTemplate { get; set; }
```

Where `FileUploadButtonTemplateContext` would be:

```csharp
public class FileUploadButtonTemplateContext
{
    public string Id { get; set; }
    public EventCallback ClearAsync { get; set; }
}
```

Or potentially, based on the existing MudBlazor `Context`-type classes for the `DataGrid` component, something like:

```csharp
public class FileUploadButtonTemplateContext<T>
{
    public string Id { get; }
    public FileUploadButtonTemplateActions Actions { get; }
    
    private readonly MudFileUpload<T> _fileUpload;
    
    public FileUploadButtonTemplateContext(MudFileUpload<T> fileUpload, string id)
    {
        _fileUpload = fileUpload;
        Id = id;
        Actions = new FileUploadButtonTemplateActions
        {
            ClearAsync = async () => await _fileUpload.ClearAsync(),
        };
    }

    public class FileUploadButtonTemplateActions
    {
        public Func<Task> ClearAsync { get; init; } = null!;
    }
}
```

Thoughts?

Mr-Technician commented:
I was thinking we would add a ClearAsync method to the file upload component and not modify the context. It could still be cleared with a button using a reference to the component. This is potentially less clean but isn't breaking.

Mr-Technician commented:
Having looked at this more closely on my computer, we might be able to get around the breaking change by add a `ToString` definition in the context class that returns the existing `for` id. This would allow most, if not all, existing code to still function as it does now.

I otherwise agree that using `ClearAsync` from a component reference is less clean. I would consider the `FileUploadButtonTemplateContext` to be the alternative to making FileUpload `Clearable` which I don't think we can do.

igotinfected committed:
WIP `ClearAsync` + `FileUploadButtonTemplateContext` implementations

igotinfected commented:
@Mr-Technician Thank you for the feedback! The `.ToString()` override is a brilliant solution, didn't think of that!

I've gone ahead and implemented the `ClearAsync` function, as well as the `FileUploadButtonTemplateContext` approach.

After adapting the snippet in the original post to use the new context, this is the result:

https://github.com/MudBlazor/MudBlazor/assets/15004223/01c45629-dea4-4123-870f-bc822e005012

Working as expected :)

For regressions, there is already a test verifying the old `context` behaviour:

https://github.com/MudBlazor/MudBlazor/blob/9cd01611e0d8185cfa6c0f34e421874f5195e066/src/MudBlazor.UnitTests/Components/FileUploadTests.cs#L108C1-L122C10

I'm also fine with going for the `ClearAsync` solution first and moving the rest to 7.0 if you're more comfortable with that. Will add more tests targeting the new context class and a doc example if we go for it.

Mr-Technician commented:
lgtm pending the new tests and docs example.

@ScarletKuro @henon Are both of you ok with merging this instead of waiting for v7? It's technically breaking but existing code will still work unless someone is doing something unusual. "Normal" usage of the button context will still work the same as it did before without modifying code due to the `ToString` method.

igotinfected committed:
Add/Update documentation and tests

igotinfected commented:
Updated all `ButtonTemplate` usages to use `@context.Id` instead of `@context`, minus the backwards compatibility test of course.

Added a render test, and two `...Context.Actions.ClearAsync` tests, one of which validates Form state.

Updated documentation and added a second drag and drop example with a form using the new `ButtonTemplate` context.

Let me know if you have different ideas for documentation, can also add a recording of the new example if you need :)

ScarletKuro commented:
>  Are both of you ok with merging this instead of waiting for v7

Personally I'm ok with this.

henon reviewed (approved):
LGTM, leaving final approval and merge to @Mr-Technician.

Mr-Technician commented:
I'll do a final review and (hopefully) merge tomorrow afternoon. 👍
